### DISCUSSION TITLE
OTAFromFile.ino example: how to generate the .bin update file using the Arduino IDE ?

### DISCUSSION THREAD

**Discussion by stefat** [earlephilhower/arduino-pico#957] [Q&A] [answered]
I've tested the OTAFromFile.ino example with an RP2040 nano connect board with success.
Now I wish to reboot my own sketch instead of the blink.bin in the example.
**My question is:how can I generate the .bin file from my own sketch using the Arduino IDE ?** 

I've tried to use the .uf2 file generated using the 'Sketch/Export compiled Binary' command but that file seems not working.
Before launching my test sketch I've uploaded the uf2 file from my sketch to the LittleFS area using the Arduino IDE with success, having renamed the file as 'update.bin'.
See please below my code used for testing [edited from the OTAFromFile.ino example]:

`
#include <PicoOTA.h>
#include <LittleFS.h>
//#include "blink_100_1000.h"

void setup() {
  Serial.begin(115200);
  delay(5000);

  Serial.println("Writing a OTA image of an <app name>.uf2 file renamed as update.bin");
  Serial.println("It is supposed to be stored in LittleFS data upload area of chip");
  LittleFS.begin();
  //File f = LittleFS.open("update.bin", "r");
  // File f = LittleFS.open("blink.bin", "w");
  // if (sizeof(blink) != f.write(blink, sizeof(blink))) {
  //   Serial.printf("Unable to write OTA binary.  Is the filesystem size set?\n");
  //   return;
  // }
  // f.close();
  // Serial.printf("done\n\n");
  if (LittleFS.exists("update.bin"))
  {
    Serial.printf("Programming OTA commands...");
    picoOTA.begin();
    picoOTA.addFile("update.bin");// renamed BLINK_MULTICORE_ST.uf2
    picoOTA.commit();
    LittleFS.end();
    Serial.printf("done\n\n");
    Serial.printf("Rebooting in 5 seconds, should begin blinker instead of this app...\n");
    delay(5000);
    rp2040.reboot();
  }
  else
  {
    LittleFS.end();
    Serial.println("Oops: update.bin file not found");
  }
}

void loop() {
}
`

The sketch output in the serial monitor is:
Writing a OTA image of an <app name>.uf2 file renamed as update.bin
It is supposed to be stored in LittleFS data upload area of chip
Programming OTA commands...done

Rebooting in 5 seconds, should begin blinker instead of this app...

But the system hangs and my sketch is not launched. I then presume the fw update went wrong and the cause is the .bin file wrong.
Any idea ?
Thank you for any help on this.

**Comment by earlephilhower** [comment]
Use the `file.bin` instead.  The UF2 version is only useful when using the USB ROM bootloader.

**Comment by stefat** [comment, marked as answer]
how can I generate a .bin instead of .uf2 ? The Sketch/Export compiler Binary command makes a .uf2 only.

**Reply by earlephilhower** [reply]
The BIN is generated on every build.  It's in the /tmp / %TEMP% folder listed.  You can turn on File->Preferences->Verbose Compiler Output to get the full path.

**Reply by stefat** [reply]
wow ! Great, I found it. Thanks !

**Reply by stefat** [reply]
BTW I'm using a MacBook Pro, macOS Ventura 13.0, M1pro, with Arduino IDE 1.8.19, and the folder in my Mac is:
/var/folders/nw/48130qln72lbtdyn1098vr0r0000gn/T/

**Comment by Andy2No** [comment]
@stefat You can set it to somewhere that's easier to remember, if you want to.  Just edit the build.path line in preferences.txt.  Mine currently looks like this:

build.path=C:\arduino-output

Set that to whatever you want, and create a folder at that location if needed.

You should find that line near the top of the file.

You can use File->Preferences to locate the preferences file, but edit and save it with Arduino closed, so it doesn't overwrite it when it exits.

**Reply by stefat** [reply]
yes ! Smart and handy suggestion. In my preferences.txt the build.path line was missing. I added that and voilà.. everything available. You saved me much time in the future. Thank you.

**Reply by Andy2No** [reply]
You're welcome, @stefat .

**Comment by timnbraun** [comment]
You can also set the config entry for sketch.always_export_binaries=true to get a 'build' directory in your sketch directory with the binaries and map file generated. I use arduino-cli, so I'm not sure where this setting is in the IDE.

**Comment by learnfromgirls** [comment]
Thanks guys for this info. I was having same problem of using the uf2 as a bin for my ota updates.
Now I have successful webserver (actually github)  based ota update on arduino nano rp2040 connect. This was a lot of work as I had to copy 
HTTPClient.h, .cpp and HTTPUpdate.h, .cpp  from the pico libraries  up to my sketch, rename them to HTTPClient2, HTTPUpdate2 and heavily edit to work with WiFiNINA_Generic. Editing was mostly deleting anything that referenced the pico-w wifi as arduino does not use the same wifi chip.
Side effect was a smaller test binary with ota.
Sketch uses 101780 bytes (1%) of program storage space. Maximum is 8384512 bytes.
Global variables use 8944 bytes (3%) of dynamic memory, leaving 253200 bytes for local variables. Maximum is 262144 bytes

**Reply by stefat** [reply]
hello, thanks for info. Do you have a public repo available ?
